### PR TITLE
Set immediate error callbacks

### PR DESCRIPTION
**Pull request by dallonf**
This is an attempt to fix https://github.com/apollographql/apollo-client/issues/1141 . Unfortunately I had a lot of trouble testing this - it's just hard to coerce Mocha to allow uncaught exceptions.

### CODE REVIEW

File: src/core/QueryManager.ts [hunk 1]
```diff
@@ -375,11 +375,8 @@ export class QueryManager {
             networkError: queryStoreValue.networkError,
           });
           if (observer.error) {
-            try {
-              observer.error(apolloError);
-            } catch (e) {
-              console.error(`Error in observer.error \n${e.stack}`);
-            }
+            // defer to avoid potential errors propagating back to Apollo
+            (setImmediate || setTimeout)(() => observer.error(apolloError));
```
Comment (calebmer) [inline src/core/QueryManager.ts]: Could we just use `setTimeout(() => { ... }, 0)` here? Is there any advantage to using `setImmediate`?

File: src/core/QueryManager.ts [hunk 2]
```diff
@@ -438,11 +435,10 @@ export class QueryManager {
 
               if (isDifferentResult) {
                 lastResult = resultFromStore;
-                try {
+                // defer to avoid potential errors propagating back to Apollo
+                (setImmediate || setTimeout)(() => {
```
Comment (calebmer) [inline src/core/QueryManager.ts]: Same here.

**Comment by calebmer** [comment]
It looks like TypeScript is failing in CI and not the actual tests. Could you try changing the [`Observer`](https://github.com/apollographql/apollo-client/blob/f60bc631bff1fce7090e7bb6fb4cc24f79878fc2/src/util/Observable.ts#L37-L41) type to:

```ts
export interface Observer<T> {
  readonly next?: (value: T) => void;
  readonly error?: (error: Error) => void;
  readonly complete?: () => void;
}
```

I _think_ I recall `readonly` fixing this error in the past for me.

**Comment by abhiaiyer91** [comment]
@calebmer tried applying your TS type fix but you still receive TS errors `error TS2532: Object is possibly 'undefined'.`?

File: src/core/QueryManager.ts (hunk 1 above)
Comment (dallonf) [inline src/core/QueryManager.ts]: `setImmediate` was called out explicitly in [this comment](https://github.com/apollographql/apollo-client/issues/1141#issuecomment-272188613), not sure if there's a real advantage, or if it's enough to offset the loss in readability. I can drop it!

**Comment by dallonf** [comment]
Went ahead and fixed the type errors. Now we're getting into the weird test stuff. I can maybe fix some of them.

**Comment by calebmer** [comment]
@abhiaiyer91 the current commit should fix the errors 😊

**Comment by dallonf** [comment]
For the `QueryManager result transformation transforms watchQuery() results` test, it's failing because the response has actually been transformed _three_ times: once in `QueryManager.ts` (line 443, when calling `observer.next()`), once in `ObservableQuery.ts`(line 198, in `refetch` after calling `queryManager.fetchQuery()`), and then again in the same place as the first time.

It was calling it three times before the changes in this PR, but the third call was made _after_ the assertion, so the test passed. Now the timing has changed subtly. I have no idea what the desired behavior is here... it's possible there was already a bug here that the tests didn't catch until now.

**Comment by dallonf** [comment]
Similarly, the `client forceFetch can temporarily be disabled with ssrForceFetchDelay` test fails because it's very dependent on the subtle timing of the observer.

I wonder if it would be a less impactful solution to do something like this...

```
try {
  observer.next(/*...*/)
} catch( err) {
  // throw error outside the current flow
 setTimeout(() => { throw err; }, 0); 
}
```

It's a lot harder to read and understand what's going on, but it would probably break fewer tests.

**Comment by calebmer** [comment]
I doubt this change actually breaks real apps. It’s unfortunate that it breaks a lot of tests 😣

**Comment by dallonf** [comment]
Yeah, had a much easier time with that approach: #1367. Let me know if I should close this one if you'd rather go with this approach and somehow fix all the tests.

**Comment by calebmer** [comment]
Yeah, https://github.com/apollographql/apollo-client/pull/1367 looks really nice. I don’t think there is any real substantial difference between the two approaches. Since the tests are green let’s pursue that one 👍

**Pull request closed without merging**